[WIP] feat(codelabs): add intermediate lit-html codelab (#982)
open-wc/open-wc: packages/codelabs/src/intermediate/lit-html.md

You can follow this codelab using anything that can display a simple HTML page. For the best editing experience, we recommend setting this up using your favorite IDE. Alternatively, you can use an [online code editor like jsbin](https://jsbin.com/?html,output).

Let's create a basic HTML page with a module script, and import LitElement from a CDN:

```html
<!DOCTYPE html>

<html>
  <body>
    <script type="module">
      import { LitElement } from 'https://unpkg.com/lit-element?module';
    </script>
  </body>
</html>
```

Make sure that you add `type="module"` to the script tag.

<aside class="notice">
In this example, we are using `unpkg`, a CDN from which we can easily import any modules that are available on NPM. When working on a real project, it is a good idea to use an actual package manager such as NPM or yarn.
</aside>

You should already know how to create a web component using `LitElement`. Go ahead and create one which renders '_My news app_' to the screen. When it works, you're ready to move on to the next step.

---

<details>
 <summary>View final result</summary>

```html
<!DOCTYPE html>
<html>
  <body>

>     <todo-app></todo-app>

Reviewer: ```suggestion
    <news-app></news-app>
```

    <script type="module">

>       import { LitElement } from 'https://unpkg.com/lit-element?module';

Reviewer: ```suggestion
      import { LitElement, html } from 'https://unpkg.com/lit-element?module';
```

      class NewsApp extends LitElement {
        render() {
          return html`
            My news app
          `;
        }
      }

      customElements.define('news-app', NewsApp);
    </script>
  </body>
</html>
```

</details>

## Google news API

For this codelab, we will be fetching live data from the google news API. You will need to retrieve an API key first:

- Go to https://newsapi.org/
- Click 'Get API key'
- Register for an account.

> - **tip**: you don't need to use an actual valid e-mail address

Reviewer: nice 😄

- Fetch articles from the news API
- Render the fetched articles as JSON

### Tips

<details>
 <summary>Using fetch</summary>
`fetch` is a browser API for making HTTP requests. It's promise based, and it returns a streaming response. We can turn the the stream into JSON using the `json` function:

```js
async fetchArticles() {
 const response = await fetch('https://newsapi.org/v2/everything?q={some-topic}&apiKey={your-api-key}');
 const jsonResponse = await response.json();
 this.articles = jsonResponse.articles;
}
```

If you're not familiar with [async await](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), this is what it looks like using `Promise`:

```js
fetchArticles() {
 fetch('https://newsapi.org/v2/everything?q={some-topic}&apiKey={your-api-key}')
 .then(response => response.json())
 .then((jsonResponse) => {
 this.articles = jsonResponse.articles;
 });
}
```

</details>

<details>
 <summary>Example connectedCallback</summary>

```js
class NewsApp extends LitElement {
  connectedCallback() {
    super.connectedCallback();

    if (!this.articles) {
      this.fetchArticles();
    }
  }
}
```

</details>

<details>

> <summary>Example render articles as JSON</summary>

Reviewer: ```suggestion
<summary>Example of articles rendered as JSON</summary>
```

          }

          return html`
            <pre>${JSON.stringify(this.articles, null, 2)}</pre>
          `;
        }
      }

      customElements.define('news-app', NewsApp);
    </script>
  </body>
</html>
```

</details>

## Displaying articles

To display individual article it's best to create a new component so that we separate the logic from the main app. This should be a plain UI component, which receives the article data as plain properties to display.

### Tasks to complete this step

- Create a `news-article` element that displays the article's title and description.
- Display a list of `news-article` elements in the `news-app`, one for each article received from the news API.

### Tips

<details>
 <summary>Create an article component</summary>

```js
class NewsArticle extends LitElement {
  static get properties() {
    return {
      title: { type: String },
      description: { type: String },
    };
  }

  render() {
    return html`
      <h3>${this.title}</h3>
      <p>${this.description}</p>
    `;
  }
}

customElements.define('news-article', NewsArticle);
```

</details>

<details>
<summary>Display a list of articles</summary>

```js
render() {

>  return html`

Reviewer: indentation

                  </li>
                `,
              )}
            </ul>
          `;
        }
      }

      customElements.define('news-app', NewsApp);

      class NewsArticle extends LitElement {
        static get properties() {
          return {
            title: { type: String },
            description: { type: String },
          };
        }

        render() {
          return html`
            <h3>${this.title}</h3>
            <p>${this.description}</p>
          `;
        }
      }

      customElements.define('news-article', NewsArticle);
    </script>
  </body>
</html>
```

</details>

## Adding a read/unread toggle

Any serious news app allows the user to track whether or not they have read an article, so let's add this capability to ours as well.

As a start, you can maintain a local property for the read/unread status in the `news-article` component. We will look into lifting this data to the parent component in the next step.

### Tasks to complete this step

- Add a property on the `news-article` component which indicates whether the user has read the article.
- Display the read/unread status in the title of each article.

> - Add a button on the `news-article` component to toggle between the read/unread status, storing this status locally.

Reviewer: ```suggestion
- Add a button in the `news-article` component to toggle between the read/unread status, storing this status locally.
```

### Tips

<details>
<summary>Conditional templating</summary>

You can conditionally render something using any valid javascript expression. For simple logic, a ternary operator is sufficient:

```js
render() {
 return html`
 <h3>${this.title} (${this.read ? 'read' : 'unread'})</h3>
 `;
}
```

> If the logic is a bit more complex, separating this into a pure function is very useful. The advantage here is that we can use regular if statement, so we don't need to squash everything into a single expression:

Reviewer: ```suggestion
If the logic is a bit more complex, separating this into a pure function is very useful. The advantage here is that we can use regular if statements, so we don't need to squash everything into a single expression:
```

```js
function readStatus(read) {
  if (read) {
    return '(read)';
  }

  return '(unread)';
}

class MyElement extends LitElement {
  render() {
    return html`
      My article (${readStatus(this.read)})
    `;
  }
}
```

</details>

<details>
 <summary>Adding event listeners</summary>

> You can add event listeners using the `@` syntax. It is just sugar for `addEventListener`:

Reviewer: ```suggestion
With lit-html, you can add event listeners using the `@` syntax, which is just syntax sugar for `addEventListener`:
```

```js
render() {
 return html`

>  <button @click=${this._onClick}></button>

Reviewer: ```suggestion
   <button @click=${this._onClick}></button>
```

              )}
            </ul>
          `;
        }
      }

      customElements.define('news-app', NewsApp);

      class NewsArticle extends LitElement {
        static get properties() {
          return {
            title: { type: String },
            description: { type: String },
            read: { type: Boolean },
          };
        }

        render() {
          return html`
            <h3>${this.title} (${this.read ? 'read' : 'unread'})</h3>
            <p>${this.description}</p>
            <button @click=${this._toggleReadStatus}>
              Mark as ${this.read ? 'unread' : 'read'}
            </button>
          `;
        }

        _toggleReadStatus() {
          this.read = !this.read;
        }
      }

      customElements.define('news-article', NewsArticle);
    </script>
  </body>
</html>
```

</details>

## Add a read/unread counter

Now that the user can mark articles as read/unread, we want to display the total amount of read and unread items in the app. This counter should be displayed in the `news-app`, but we're storing the read/unread status in the `news-article` component. We need to think of a better way to solve this...

> It's best to keep the data in your application flowing in one direction from to bottom. Parent components are responsible for data of child components, including changing this data.

Reviewer: ```suggestion
It's best to keep the data in your application flowing in one direction from top to bottom. Parent components are responsible for data of child components, including changing this data.
```

- Display the total amount of read and unread articles in the `news-app` component.

### Tips

Remember that with `LitElement`, you need to use immutable data patterns. Otherwise, it will not be able to pick up data changes.

<details>
 <summary>Firing events</summary>

```js
_toggleReadStatus() {
 this.dispatchEvent(new CustomEvent('toggle-read-status'));
}
```

</details>

<details>
 <summary>Catch event from a list of elements</summary>

When you add an event listener on an element in a list of templates, you need a way to know which element in the list fired the event. This can be done by passing the list item to the event handler:

```js
html`
  ${this.articles.map(
    article => html`
      <li>
        <news-article @toggle-read-status=${() => this._toggleReadStatus(article)}></news-article>
      </li>
    `,
  )}
`;
```

</details>

<details>
 <summary>Update read status in main app</summary>

To update the read status, we need to use immutable data update patterns. This means we should create a articles array, and a new object for the article that was updated. A quick way to do this, is by using a map function:

```js

> _toggleReadStatus(articleToUpdate) {

Reviewer: indentation
